jdk/test/java/security/SignedJar/spi-calendar-provider/TestSPISigned.java

```diff
@@ -0,0 +1,136 @@
+    private static final Path PROVIDER_PARENT = Paths.get(TEST_CLASSES);
+    private static final Path PROVIDER_DIR = PROVIDER_PARENT.resolve("provider");
+    private static final Path META_INF_DIR_TARGET = PROVIDER_DIR.resolve("META-INF");
+    private static final Path SERVICES_DIR_TARGET = META_INF_DIR_TARGET.resolve("services");
+    private static final Path PROVIDER_TARGET = SERVICES_DIR_TARGET.resolve("java.util.spi.CalendarDataProvider");
+    private static final Path MODS_DIR = Paths.get("mods");
+    private static final Path UNSIGNED_JAR = MODS_DIR.resolve("unsigned-with-locale.jar");
+    private static final Path SIGNED_JAR = MODS_DIR.resolve("signed-with-locale.jar");
+
+    public static void main(String[] args) throws Throwable {
+        if (args.length == 1) {
+            String arg = args[0];
+            if ("run-test".equals(arg)) {
+                System.out.println("Debug: Running test");
+                String provProp = System.getProperty("java.ext.dirs");
+                String[] extDirs = provProp.split(File.pathSeparator);
+                boolean extDirFound = false;
+                for (int i = 0; i < extDirs.length; i++) {
+                    if (extDirs[i].equals(MODS_DIR.toAbsolutePath().toString())) {
+                        extDirFound = true;
+                        break;
+                    }
+                }
+                if (!extDirFound) {
+                   throw new RuntimeException("Test failed! Expected -Djava.ext.dirs to include CalendarDataProvider jar");
+                }
+                doRunTest();
+            } else {
+               throw new RuntimeException("Test failed! Expected 'run-test' arg for test run");
+            }
+        } else {
+            // Set up signed jar with custom calendar data provider
+            //
+            // 1. Create jar with custom CalendarDataProvider
+            Files.createDirectory(META_INF_DIR_TARGET);
+            Files.createDirectory(SERVICES_DIR_TARGET);
+            Files.copy(PROVIDER_SOURCE, PROVIDER_TARGET);
+            JarUtils.createJarFile(UNSIGNED_JAR, PROVIDER_DIR);
+
+            // create signer's keypair
+            SecurityTools.keytool("-genkeypair -keyalg RSA -keystore ks " +
+                                  "-storepass changeit -dname CN=test -alias test")
+                     .shouldHaveExitValue(0);
+            // sign jar
+            SecurityTools.jarsigner("-keystore ks -storepass changeit " +
+                                "-signedjar " + SIGNED_JAR + " " + UNSIGNED_JAR + " test")
+                     .shouldHaveExitValue(0);
+            // run test, which must not throw a NPE
+            List<String> testRun = new ArrayList<>();
+            String extDir = System.getProperty("java.home") + File.separator + "lib" + File.separator + "ext";
+            testRun.add("-Djava.ext.dirs=" + System.getProperty("java.home") + extDir + File.pathSeparator + MODS_DIR.toAbsolutePath().toString());
```

Comment:
Doesn't this add `System.getProperty("java.home")` to the front of `extDir` which already begins with `System.getProperty("java.home")`?

Reply:
If you override the property, it will only contain what you pass it:

```
$ /usr/lib/jvm/java-1.8.0-openjdk/bin/java -Djava.ext.dirs=/foo -XshowSettings:properties -version 2>&1 | grep java.ext.dirs
    java.ext.dirs = /foo
$ /usr/lib/jvm/java-1.8.0-openjdk/bin/java -XshowSettings:properties -version 2>&1 | grep java.ext.dirs
    java.ext.dirs = /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.362.b09-2.fc37.x86_64/jre/lib/ext
```

The intention is to use the extra dir as well, but it should also use `${java.home}/ext`

Reply:
Yeah, I get that, but I'm talking about the simpler issue of including the value of `System.getProperty("java.home")` twice, once in `extDir` and then again in the string passed to `testRun.add`.

If I break out the code into a test file,

~~~
$ cat TestExt.java
import java.io.File;

import java.nio.file.Path;
import java.nio.file.Paths;

import java.util.ArrayList;
import java.util.List;

public class TestExt
{
    private static final Path MODS_DIR = Paths.get("mods");

    public static void main(String[] args) {
	// run test, which must not throw a NPE
	List<String> testRun = new ArrayList<>();
	String extDir = System.getProperty("java.home") + File.separator + "lib" + File.separator + "ext";
	testRun.add("-Djava.ext.dirs=" + System.getProperty("java.home") + extDir + File.pathSeparator + MODS_DIR.toAbsolutePath().toString());
	System.out.println(testRun);
    }
}
$ $HOME/build/icedtea8/bin/java -version
openjdk version "1.8.0_362"
OpenJDK Runtime Environment (IcedTea 3.26.0) (Gentoo build 1.8.0_362-b09)
OpenJDK 64-Bit Server VM (build 25.362-b09, mixed mode)
$ $HOME/build/icedtea8/bin/javac TestExt.java 
$ $HOME/build/icedtea8/bin/java TestExt
[-Djava.ext.dirs=/home/andrew/build/icedtea8/jre/home/andrew/build/icedtea8/jre/lib/ext:/tmp/mods]
~~~

You can see the value passed includes `java.home` twice. I guess it's not causing an issue with the test because it doesn't actually need to use any of the extensions in the system extension directory.